File: backend/app/schemas/message.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class MessageCreate(BaseModel):
    content: str
    recipient_id: int
    listing_id: Optional[int] = None


class MessageResponse(BaseModel):
    id: int
    conversation_id: int
    sender_id: int
    sender_name: str
    content: str
    is_read: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ConversationResponse(BaseModel):
    id: int
    user1_id: int
    user2_id: int
    listing_id: Optional[int] = None
    listing_title: Optional[str] = None
    other_user_id: int
    other_user_name: str
    other_user_email: Optional[str] = None
    last_message: Optional[MessageResponse] = None
    unread_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ConversationDetailResponse(BaseModel):
    id: int
    user1_id: int
    user2_id: int
    listing_id: Optional[int] = None
    listing_title: Optional[str] = None
    other_user_id: int
    other_user_name: str
    other_user_email: Optional[str] = None
    messages: List[MessageResponse]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: packages/shared/src/schemas/message.ts
import { z } from "zod";

export const MessageCreateSchema = z.object({
  recipient_id: z.number(),
  listing_id: z.number().optional(),
});

export const MessageResponseSchema = z.object({
  id: z.number(),
  conversation_id: z.number(),
  sender_id: z.number(),
  sender_name: z.string(),
  content: z.string(),
  created_at: z.string().datetime(),
});

export const ConversationResponseSchema = z.object({
  id: z.number(),
  user1_id: z.number(),
  user2_id: z.number(),
  listing_id: z.number().nullable().optional(),
  listing_title: z.string().nullable().optional(),
  other_user_id: z.number(),
  other_user_name: z.string(),
  other_user_email: z.string().nullable().optional(),
  last_message: MessageResponseSchema.nullable().optional(),
  unread_count: z.number(),
  created_at: z.string().datetime(),
});

export const ConversationDetailResponseSchema = z.object({
  id: z.number(),
  user1_id: z.number(),
  user2_id: z.number(),
  listing_id: z.number().nullable().optional(),
  listing_title: z.string().nullable().optional(),
  other_user_id: z.number(),
  other_user_name: z.string(),
  other_user_email: z.string().nullable().optional(),
  messages: z.array(MessageResponseSchema),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export type MessageCreate = z.infer<typeof MessageCreateSchema>;
export type MessageResponse = z.infer<typeof MessageResponseSchema>;
export type ConversationResponse = z.infer<typeof ConversationResponseSchema>;
export type ConversationDetailResponse = z.infer<typeof ConversationDetailResponseSchema>;

